torch/distributed/tensor/_collective_utils.py

```diff
@@ -31,13 +29,10 @@
     @torch.library.register_fake("_dtensor::shard_dim_alltoall")
     def _shard_dim_alltoall_meta(input, gather_dim, shard_dim, group_name):
         group_size = _get_group_size_by_name(group_name)
-        stacked_list = [torch.empty_like(input) for _ in range(group_size)]
-        group = _resolve_process_group(group_name)
-        group_rank = get_group_rank(group, get_rank())
-
-        return torch.cat(stacked_list, dim=gather_dim).chunk(group_size, dim=shard_dim)[
-            group_rank
-        ]
+        shape = list(input.shape)
+        shape[gather_dim] *= group_size
+        shape[shard_dim] //= group_size
+        return torch.empty(shape, device=input.device, dtype=input.dtype)
```

Comment:
the process group logic may have something to do with uneven sharding? cc: @wz337 
Can we just replicate what the actual implementation of `shard_dim_alltoall` does:
```
return out.contiguous() if not out.is_contiguous() else out
```
https://github.com/pytorch/pytorch/pull/141835/files#diff-61ee7944e7972ab91d3d0d347cb87b8161db88e58ea69b727f12afd0d49d622eL65

Reply:
I think Tianyu's suggestion looks good.

Curious if it is possible to have different `input.shape` across ranks. If so the original code makes sense otherwise the new change is good.

Reply:
Reverted appropriately to just call `contiguous()` with the original impl.

I've excluded the ternary as both the [documentation](https://pytorch.org/docs/stable/generated/torch.Tensor.contiguous.html) and the [code I can find](https://github.com/pytorch/pytorch/blob/4959784dac633cc23fc59401eb8cfcae47a07f4c/aten/src/ATen/native/TensorProperties.cpp#L108) appears to indicate `t.contiguous()` checks `is_contiguous()` internally before attempting a copy. please correct if mistaken.

Reply:
Sounds good. Would you please also remove the `is_contiguous` check in the original impl :-)

Reply:
:+1: